lib/staccato/v4/tracker.rb

```diff
@@ -1,155 +1,334 @@
+    #     client
+    # @param user_id [String, nil] A unique identifier for a user
+    # @param user_properties [Hash] custom user properties that describe
+    #     segments of your user base, such as language preference or geographic
+    #     location
+    # @param non_personalized_ads [Boolean] Set to true to indicate these events
+    #     should not be used for personalized ads
+    # @param time [Time] The time to associate with the event.  This should
+    #     only be set to record events that happened in the past.  This value
+    #     can be overridden via user_property or event timestamps.
+    # @param default_event_params [Hash] default event params used by all events
+    #     sent from this tracker
+    def initialize(measurement_id, api_secret, client_id,
+                   user_id: nil, user_properties: nil,
+                   non_personalized_ads: nil,
+                   time: nil, timestamp_micros: nil,
+                   default_event_params: {})
+      raise ArgumentError, 'measurement_id is required' unless measurement_id
+      raise ArgumentError, 'api_secret is required'     unless api_secret
+      raise ArgumentError, 'client_id is required'      unless client_id
+      if time && timestamp_micros
+        raise ArgumentError, 'time and timestamp_micros are mutually exclusive'
+      end
+
+      @measurement_id       = measurement_id
+      @api_secret           = api_secret
+      @client_id            = client_id
+      @user_id              = user_id&.to_s
+      @user_properties      = user_properties&.to_hash
+      @default_event_params = default_event_params.to_hash
+      @time                 = time&.to_time
+      @timestamp_micros     = Integer(timestamp_micros) if timestamp_micros
 
-      self.events = []
+      if @timestamp_micros then @time = Time.at(@timestamp_micros / 1_000_000)
+      elsif @time then @timestamp_micros = Staccato::V4.timestamp_micros(@time)
+      end
+
+      @adapters = []
+      @events   = []
     end
 
     def adapter=(adapter)
       @adapters = [adapter]
     end
 
-    def add_adapter(adapter)
+    # Adds an adapter to the Tracker.  If no adapter was explicitly added, this
+    # will be the only adapter for the tracker.
+    #
+    # To add an adapter in addition to the {.default_adapter}, add it directly
+    # to the adapters array:
+    #
+    #    tracker.adapters << Staccato::Adapter::Logger.new
+    #
+    # @param adapter [Object] an adapter object, or use one of the following
+    #     symbols to configure a default adapter: `:net_http`, `:validation`,
+    #     `:logger`.
+    # @param logger [Logger] used when `adapter` is `:logger`
+    # @param formatter [Proc] used when `adapter` is `:logger`
+    def add_adapter(adapter, logger = nil, formatter = nil)
+      case adapter
+      when :logger
+        require 'staccato/adapter/logger'
+        uri = self.class.collection_uri
+        adapter = Staccato::Adapter::Logger.new(uri, logger, formatter)
```

Comment:
Could this follow the same pattern as the other adapters in the defaults module? `self.class.logging_adapter(uri, logger, formatter)`?

Reply:
I considered that.  I decided against it because those were effectively accessors for shared singleton-ish objects, and this would need to generate a new logger adapter every time.